File: frontend/src/modules/floor/geometry.ts
/**
 * Where the chairs go.
 *
 * The floor view draws the room rather than a grid of status squares, which
 * means it has to decide, for a given table, where each seat physically sits.
 * That is arithmetic, not presentation, so it lives here with tests rather than
 * inside a component where it cannot be checked.
 *
 * The rules are the ones a person laying out furniture would use:
 *
 *   * **Round tables** space their chairs evenly around the circle.
 *   * **Rectangles** put chairs on the long sides first and only use the ends
 *     when the sides are full — six chairs around a 2×1 table is 3+3, not
 *     2+2+1+1, because nobody seats two people at the end of a narrow table
 *     while the sides have room.
 *   * **Booths** are against a wall: chairs on three sides, never the back.
 *   * **Bars** are a single row facing the counter.
 *
 * Occupancy fills seats in order rather than at random. A party of two at a
 * six-top sits together, and a diagram that scattered them would be drawing
 * something that does not happen.
 */

export type TableShape = 'ROUND' | 'SQUARE' | 'RECT' | 'BOOTH' | 'BAR'

export interface Seat {
  /** Fraction of the table's own width, measured from its centre. */
  x: number
  /** Fraction of the table's own height, measured from its centre. */
  y: number
  /** Degrees the chair is turned, so its back faces away from the table. */
  angle: number
  occupied: boolean
}

/** How far outside the table edge a chair sits, as a fraction of the table. */
const GAP = 0.78

function ring(count: number, occupied: number): Seat[] {
  return Array.from({ length: count }, (_, index) => {
    // Start at the top and go clockwise, so seat 1 is where a person would say
    // "the one facing the door" rather than an arbitrary point.
    const angle = (index / count) * 360 - 90
    const radians = (angle * Math.PI) / 180
    return {
      x: Math.cos(radians) * GAP,
      y: Math.sin(radians) * GAP,
      angle: angle + 90,
      occupied: index < occupied,
    }
  })
}

/**
 * How many chairs fit along one edge that is `span` grid cells long.
 *
 * A one-cell edge takes two people; each extra cell adds one. That is not a
 * formula from anywhere — it is what a 60 cm chair against a 70 cm table edge
 * comes to, and it produces the layouts people actually recognise: four around
 * a small square, six around a 2×1.
 */
function edgeCapacity(span: number): number {
  return Math.max(2, span + 1)
}

/**
 * Distribute `count` chairs around a rectangle, long sides first.
 *
 * The ends are used only once the long sides are full, because nobody seats two
 * people at the narrow end of a table while the sides still have room. Six
 * around a 2×1 is 3+3, not 2+2+1+1.
 */
function perimeter(count: number, spanX: number, spanY: number, occupied: number): Seat[] {
  const horizontalIsLong = spanX >= spanY
  const longSpan = horizontalIsLong ? spanX : spanY
  const shortSpan = horizontalIsLong ? spanY : spanX

  const perLongSide = edgeCapacity(longSpan)
  const onSides = Math.min(count, perLongSide * 2)
  const firstSide = Math.ceil(onSides / 2)
  const secondSide = onSides - firstSide

  // Whatever will not fit on the sides goes to the ends, split evenly and
  // capped by how wide the end actually is.
  const overflow = count - onSides
  const perEndCap = edgeCapacity(shortSpan)
  const firstEnd = Math.min(perEndCap, Math.ceil(overflow / 2))
  const secondEnd = Math.min(perEndCap, overflow - firstEnd)

  const seats: Seat[] = []
  const place = (n: number, side: 'top' | 'bottom' | 'left' | 'right') => {
    for (let index = 0; index < n; index += 1) {
      // Evenly spaced along the side, inset from the corners.
      const t = (index + 1) / (n + 1) - 0.5
      if (side === 'top') seats.push({ x: t * 1.5, y: -GAP, angle: 180, occupied: false })
      if (side === 'bottom') seats.push({ x: t * 1.5, y: GAP, angle: 0, occupied: false })
      if (side === 'left') seats.push({ x: -GAP, y: t * 1.5, angle: 90, occupied: false })
      if (side === 'right') seats.push({ x: GAP, y: t * 1.5, angle: 270, occupied: false })
    }
  }

  if (horizontalIsLong) {
    place(firstSide, 'top')
    place(secondSide, 'bottom')
    place(firstEnd, 'left')
    place(secondEnd, 'right')
  } else {
    place(firstSide, 'left')
    place(secondSide, 'right')
    place(firstEnd, 'top')
    place(secondEnd, 'bottom')
  }

  return seats.map((seat, index) => ({ ...seat, occupied: index < occupied }))
}

export function seatsFor(
  shape: TableShape,
  count: number,
  occupied = 0,
  spanX = 1,
  spanY = 1,
): Seat[] {
  const seated = Math.max(0, Math.min(occupied, count))
  if (count <= 0) return []

  switch (shape) {
    case 'ROUND':
      return ring(count, seated)

    case 'BAR':
      // One row facing the counter. A bar with chairs behind it is a table.
      return Array.from({ length: count }, (_, index) => ({
        x: ((index + 1) / (count + 1) - 0.5) * 1.7,
        y: GAP,
        angle: 0,
        occupied: index < seated,
      }))

    case 'BOOTH': {
      // Against a wall: three sides, never the back.
      const perSide = Math.ceil(count / 3)
      const seats: Seat[] = []
      for (let index = 0; index < count; index += 1) {
        const side = Math.floor(index / perSide)
        const withinSide = index % perSide
        const t = (withinSide + 1) / (perSide + 1) - 0.5
        if (side === 0) seats.push({ x: t * 1.5, y: GAP, angle: 0, occupied: false })
        else if (side === 1) seats.push({ x: -GAP, y: t * 1.5, angle: 90, occupied: false })
        else seats.push({ x: GAP, y: t * 1.5, angle: 270, occupied: false })
      }
      return seats.map((seat, index) => ({ ...seat, occupied: index < seated }))
    }

    case 'RECT':
    case 'SQUARE':
    default:
      return perimeter(count, spanX, spanY, seated)
  }
}

/** Pixel footprint of a table, before rotation. */
export function footprint(shape: TableShape, spanX: number, spanY: number, cell: number) {
  const width = spanX * cell
  const height = spanY * cell
  // A bar is a counter: long and shallow, whatever its span says.
  if (shape === 'BAR') return { width, height: Math.round(cell * 0.42) }
  return { width, height }
}

/**
 * Split the seats into the ones behind the table and the ones in front.
 *
 * The room is drawn tilted, so a chair at the bottom of a table is NEARER the
 * viewer and must be painted over the table top, while one at the top is behind
 * it and must be painted under. Drawing them all in one pass — which is what
 * the first version did — puts the near chairs behind the furniture and the
 * whole scene stops reading as three-dimensional.
 */
export function splitByDepth(seats: Seat[]): { behind: Seat[]; infront: Seat[] } {
  return {
    behind: seats.filter((seat) => seat.y < 0),
    infront: seats.filter((seat) => seat.y >= 0),
  }
}

export type Fullness = 'free' | 'light' | 'busy' | 'full'

/**
 * How full a table is, as a word.
 *
 * Used for colour AND stated in the label. "4 seats, 2 seated" is the thing a
 * waiter seating a walk-in actually needs; a table that merely reads "occupied"
 * hides four empty chairs.
 */
export function fullness(seats: number, seated: number): Fullness {
  if (seated <= 0) return 'free'
  if (seated >= seats) return 'full'
  return seated / seats >= 0.6 ? 'busy' : 'light'
}

File: desktop/src/caesar_pos/ui/floor/geometry.py
"""
Where the chairs go — the Desktop's copy of the Web's seat arithmetic.

Deliberately a copy rather than a shared module: the Web's version is
TypeScript running in a browser and this one is Python driving QPainter, so
there is no artefact to vendor. What IS shared is the behaviour, and
`tests/test_floor_geometry.py` asserts the two agree case for case — the same
discipline the money modules follow, enforced by the same kind of test.

The rules are the ones a person laying out furniture would use:

  * **Round tables** space their chairs evenly around the circle.
  * **Rectangles** put chairs on the long sides first and use the ends only when
    the sides are full. Six around a 2×1 table is 3+3, not 2+2+1+1, because
    nobody seats two people at the narrow end while the sides have room.
  * **Booths** are against a wall: three sides, never the back.
  * **Bars** are one row facing the counter.

Occupancy fills seats in order. A party of two at a six-top sits together, and
scattering them would draw something that does not happen.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

ROUND = "ROUND"
SQUARE = "SQUARE"
RECT = "RECT"
BOOTH = "BOOTH"
BAR = "BAR"

#: How far outside the table edge a chair sits, as a fraction of the table.
GAP = 0.78


@dataclass(frozen=True)
class Seat:
    """Position as a fraction of the table's own size, measured from its centre."""

    x: float
    y: float
    angle: float
    occupied: bool


def edge_capacity(span: int) -> int:
    """
    How many chairs fit along an edge `span` grid cells long.

    A one-cell edge takes two people; each extra cell adds one. Not a formula
    from anywhere — it is what a 60cm chair against a 70cm table edge comes to,
    and it produces the layouts people recognise.
    """
    return max(2, span + 1)


def _ring(count: int, occupied: int) -> list[Seat]:
    seats = []
    for index in range(count):
        # Start at the top and go clockwise, so seat one is the one a person
        # would call "facing the door" rather than an arbitrary point.
        angle = (index / count) * 360 - 90
        radians = math.radians(angle)
        seats.append(
            Seat(
                x=math.cos(radians) * GAP,
                y=math.sin(radians) * GAP,
                angle=angle + 90,
                occupied=index < occupied,
            )
        )
    return seats


def _perimeter(count: int, span_x: int, span_y: int, occupied: int) -> list[Seat]:
    horizontal_is_long = span_x >= span_y
    long_span = span_x if horizontal_is_long else span_y
    short_span = span_y if horizontal_is_long else span_x

    per_long_side = edge_capacity(long_span)
    on_sides = min(count, per_long_side * 2)
    first_side = math.ceil(on_sides / 2)
    second_side = on_sides - first_side

    overflow = count - on_sides
    per_end_cap = edge_capacity(short_span)
    first_end = min(per_end_cap, math.ceil(overflow / 2))
    second_end = min(per_end_cap, overflow - first_end)

    seats: list[Seat] = []

    def place(n: int, side: str) -> None:
        for index in range(n):
            t = (index + 1) / (n + 1) - 0.5
            if side == "top":
                seats.append(Seat(t * 1.5, -GAP, 180, False))
            elif side == "bottom":
                seats.append(Seat(t * 1.5, GAP, 0, False))
            elif side == "left":
                seats.append(Seat(-GAP, t * 1.5, 90, False))
            else:
                seats.append(Seat(GAP, t * 1.5, 270, False))

    if horizontal_is_long:
        place(first_side, "top")
        place(second_side, "bottom")
        place(first_end, "left")
        place(second_end, "right")
    else:
        place(first_side, "left")
        place(second_side, "right")
        place(first_end, "top")
        place(second_end, "bottom")

    return [Seat(seat.x, seat.y, seat.angle, index < occupied) for index, seat in enumerate(seats)]


def seats_for(
    shape: str, count: int, occupied: int = 0, span_x: int = 1, span_y: int = 1
) -> list[Seat]:
    if count <= 0:
        return []
    seated = max(0, min(occupied, count))

    if shape == ROUND:
        return _ring(count, seated)

    if shape == BAR:
        # One row facing the counter. A bar with chairs behind it is a table.
        return [
            Seat(((index + 1) / (count + 1) - 0.5) * 1.7, GAP, 0, index < seated)
            for index in range(count)
        ]

    if shape == BOOTH:
        # Against a wall: three sides, never the back.
        per_side = math.ceil(count / 3)
        seats: list[Seat] = []
        for index in range(count):
            side = index // per_side
            t = (index % per_side + 1) / (per_side + 1) - 0.5
            if side == 0:
                seats.append(Seat(t * 1.5, GAP, 0, False))
            elif side == 1:
                seats.append(Seat(-GAP, t * 1.5, 90, False))
            else:
                seats.append(Seat(GAP, t * 1.5, 270, False))
        return [Seat(s.x, s.y, s.angle, i < seated) for i, s in enumerate(seats)]

    return _perimeter(count, span_x, span_y, seated)


def footprint(shape: str, span_x: int, span_y: int, cell: int) -> tuple[int, int]:
    """Pixel size of a table before rotation."""
    width = span_x * cell
    height = span_y * cell
    if shape == BAR:
        # A bar is a counter: long and shallow, whatever its span says.
        return width, round(cell * 0.42)
    return width, height


FREE = "free"
LIGHT = "light"
BUSY = "busy"
FULL = "full"

STATE_LABELS = {
    FREE: "متاحة",
    LIGHT: "جالسون قليل",
    BUSY: "شبه ممتلئة",
    FULL: "ممتلئة",
}


def fullness(seats: int, seated: int) -> str:
    """
    How full a table is, as a word.

    Used for colour AND said in the label. "4 كراسي، 2 جالسين" is what a waiter
    seating a walk-in needs; a table that merely reads "occupied" hides two
    empty chairs.
    """
    if seated <= 0:
        return FREE
    if seated >= seats:
        return FULL
    return BUSY if seated / seats >= 0.6 else LIGHT
